src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -19,8 +19,15 @@ namespace gcs {
 
 double LeastResourceScorer::Score(const ResourceSet &required_resources,
                                   const SchedulingResources &node_resources) {
-  const auto &available_resources = node_resources.GetAvailableResources();
-  const auto &available_resource_amount_map = available_resources.GetResourceAmountMap();
+  ResourceSet new_available_resource_set;
```

Comment:
Could you add some comments here about why you do it in this way?

Reply:
First, we make `available_resource_set` a pointer to avoid copy overhead in most cases (there is no normal task and we would not have to get into the `if` branch).

Second, if we get into the `if` branch, we may define `new_available_resource_set` there and make `available_resource_set` point to it. But after the `if` branch, the `new_available_resource_set` would be destructed (because its lifetime is within the `if` scope) and `available_resource_set` will point to a un-initialized space. So we define the `new_available_resource_set` before of the `if` scope.

These are some trivial details, and I'm not sure if it can be added there as comments.

Reply:
Sorry, I think I didn't make it clear. My question is that:

why previous we use `node_resources.GetAvailableResources` directly, and now we copy it back and use the new_available_resource_set instead of operating it directly in node_resources.

Reply:
In GCS scheduling, we want the `resources_available_` (of `class SchedulingResources`) only acquired or released by actor scheduling, instead of being updated by resource reports from raylets (See the new `UpdateFromResourceReport` function). So the 'actual' available resource (if there exist normal tasks) is equal to 'resources_available_' - `resources_normal_tasks_`.